wasmtime-environ/src/cache.rs

```diff
@@ -505,3 +505,343 @@ impl<'de> Deserialize<'de> for JtOffsetsWrapper<'_> {
+                                    end: rng.gen(),
+                                })
+                                .collect(),
+                        )
+                    })
+                    .collect()
+            })
+            .collect();
+
+        let stack_slots = (0..rng.gen_range(0, 0x6))
+            .map(|_| {
+                let mut slots = ir::StackSlots::new();
+                slots.push(ir::StackSlotData {
+                    kind: ir::StackSlotKind::SpillSlot,
+                    size: rng.gen(),
+                    offset: rng.gen(),
+                });
+                slots.frame_size = rng.gen();
+                slots
+            })
+            .collect();
+
+        ModuleCacheData::from_tuple((
+            Compilation::new(funcs),
+            relocs,
+            trans,
+            value_ranges,
+            stack_slots,
+        ))
+    }
+
+    impl ModuleCacheEntry {
+        pub fn mod_cache_path(&self) -> &Option<PathBuf> {
+            &self.mod_cache_path
+        }
+    }
+
+    // cranelift's types (including SecondaryMap) doesn't implement PartialEq
+    impl PartialEq for ModuleCacheData {
+        fn eq(&self, other: &Self) -> bool {
+            // compilation field
+            if self.compilation.len() != other.compilation.len() {
+                return false;
+            }
+
+            for i in (0..self.compilation.len()).map(DefinedFuncIndex::new) {
+                let lhs = self.compilation.get(i);
+                let rhs = other.compilation.get(i);
+                if lhs.body != rhs.body
+                    || lhs.jt_offsets.get_default() != rhs.jt_offsets.get_default()
+                {
+                    return false;
+                }
+
+                for j in (0..max(lhs.jt_offsets.len(), rhs.jt_offsets.len())).map(JumpTable::new) {
+                    if lhs.jt_offsets.get(j) != rhs.jt_offsets.get(j) {
+                        return false;
+                    }
+                }
+            }
+
+            // relocs
+            {
+                if self.relocations.len() != other.relocations.len() {
+                    return false;
+                }
+                let it_lhs = self.relocations.iter();
+                let it_rhs = other.relocations.iter();
+                if it_lhs.zip(it_rhs).any(|(lhs, rhs)| lhs != rhs) {
+                    return false;
+                }
+            }
+
+            // debug symbols
+            {
+                if self.address_transforms.len() != other.address_transforms.len() {
+                    return false;
+                }
+                let it_lhs = self.address_transforms.iter();
+                let it_rhs = other.address_transforms.iter();
+                if it_lhs.zip(it_rhs).any(|(lhs, rhs)| lhs != rhs) {
+                    return false;
+                }
+            }
+
+            true
+        }
+    }
+
+    // binemit::Reloc doesn't implement PartialEq
```

Comment:
We should make `binemit::Reloc` implement `PartialEq` and `Eq`.

Reply:
Same thing as with `SecondaryMap` - I'll do it later as follow up refactor.